src/main/flight/autopilot.c

```diff
@@ -91,17 +206,182 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+                updateLocation(efAxis, loopAxis);
+                // while sticks are moving, reset the location on each axis, to maintain a usable D value
+                // slowly leak iTerm away, approx 2s time constant
+                efAxis->integral *= leak;
+                // increase sanity check distance depending on speed, typically maximal when sticks stop
+            } else if (efAxis->isStopping) {
+                // 'phase' after sticks stop, but before craft has stopped
+                pidD *= 1.6f; // aribitrary D boost to stop more quickly when sticks are centered
+                if (velocity * velocityFiltered < 0.0f) {
+                    // when an axis has nearly stopped moving, reset it and end it's start phase
+                    updateLocation(efAxis, loopAxis);
+                    efAxis->isStopping = false;
+                }
+            }
+
+            float pidDA = pidD + pidA;
+
+            // limit sum of D and A per axis based on total DA vector length, otherwise can be too aggressive when starting at speed
+            // limit is 35 degrees from D and A alone, arbitrary value.  20 is a bit too low, allows a lot of overshoot
+            // note: an angle of more than 35 degrees can still be achieved as P and I grow
+            const float pidDASquared = sq(pidD + pidA);
+            float magSq = pidDASquared + prevPidDASquared;
+            if (magSq > sqMaxDAAngle && magSq > 0.0f) {
+                pidDA *= maxDAAngle / sqrtf(magSq);
+            }
+            prevPidDASquared = pidDASquared;
+
+            // ** PID Sum **
+            efAxis->pidSum = pidP + pidI + pidDA;
+
+            if (gyroConfig()->gyro_filter_debug_axis == loopAxis) {
+                DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 0, lrintf(distanceCm));
+                DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 4, lrintf(pidP * 10));
+                DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 5, lrintf(pidI * 10));
+                DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 6, lrintf(pidD * 10));
+                DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 7, lrintf(pidA * 10));
+            }
+        } // end for loop
+
+        if (ap.sticksActive) {
+            // keep update sanity check distance while sticks are out
+            ap.sanityCheckDistance = gpsSol.groundSpeed > 500 ? gpsSol.groundSpeed * 2.0f : 1000.0f;
+            // if a Position Hold deadband is set, and sticks are outside deadband, allow pilot control in angle mode
+            ap.pidSumCraft[AI_ROLL] = 0.0f;
+            ap.pidSumCraft[AI_PITCH] = 0.0f;
+        } else {
+            // ** Rotate pid Sum to quad frame of reference, into pitch and roll **
+            const float headingRads = DECIDEGREES_TO_RADIANS(attitude.values.yaw);
+            const float sinHeading = sin_approx(headingRads);
+            const float cosHeading = cos_approx(headingRads);
+            ap.pidSumCraft[AI_ROLL] = -sinHeading * ap.efAxis[NS].pidSum + cosHeading * ap.efAxis[EW].pidSum;
```

Comment:
It did take a long time, because I had no idea how to do the transformation, and there was nothing I could find that would do it for me.  And there are conventions to the directions / signs of movements for each axis, but these are not well documented.

If we had a 'standard' way of doing this, it would have saved me a lot of time.

It would be great if we had a function in Betaflight that made such a transform possible in the 'standard' way.  Maybe there is?  I don't know?  No-one pointed me to existing code that converts earth frame values in latitude and longitude amounts into pitch and roll values.  I had to figure out something that worked all by myself.  And yes, it was difficult, and took a long time.

For now, this works, anyone can see how it can to be done using simple sine and cosine maths.

If you have a better or more effective way of doing this, perhaps using some of the imu code, or if you would like to propose a better way, that would be great.

Reply:
Most code is in https://github.com/ctzsnooze/betaflight/pull/24, but I'll check the rotations tomorrow.

One problem is that this rotation is actually 2 step process. First you convert EF coordinates into body frame (good choice for ENU is X forward and Y left (and Z up), right handed; this it is all mixed up in other code, at least some confusion is my fault). Then you can convert desired axis velocities into angles - positive Y rotation (pitch) to fly forward and negative X rotation (roll) to fly left. That is also 90 deg rotation.
You code combines both into single equation (together with heading zero reference and heading angle direction). The code does work and can be decomposed into respective parts. But it takes time (and a lot of paper).

We should define relevant coordinate systems, ideally with interactive visualization, Use something close to existing system (for example ENU, XYZ as described above, but it may not match IMU and rotation matrix) and adapt existing code to match.

Or use coordinate system from PX4 or Ardupilot (IIRC they don't match) and update BF code to match.